internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,637 @@
+	}
+	if len(dbClusterOutput.DBClusters) != 1 {
+		return nil, fmt.Errorf("expected exactly one RDS cluster for installation id %s (found %d)", d.installationID, len(dbClusterOutput.DBClusters))
+	}
+
+	return dbClusterOutput.DBClusters[0], nil
+}
+
+func (d *RDSMultitenantDatabase) acquireLock(dbClusterID string, store model.InstallationDatabaseStoreInterface) (func(logger log.FieldLogger), error) {
+	locked, err := store.LockDatabaseCluster(dbClusterID, d.installationID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not acquire lock for database cluster %s", dbClusterID)
+	}
+	if !locked {
+		return nil, errors.Errorf("could not acquire lock for database cluster %s", dbClusterID)
+	}
+	unlockFN := func(logger log.FieldLogger) {
+		unlocked, err := store.UnlockDatabaseCluster(dbClusterID, d.installationID, true)
+		if err != nil {
+			logger.WithError(err).Errorf("provisioner datastore failed to release locker for database cluster ID %s", dbClusterID)
+		}
+		if !unlocked {
+			logger.Warnf("database ID %s and locker ID %s is still locked", dbClusterID, d.installationID)
+		}
+	}
+
+	return unlockFN, nil
+}
+
+func (d *RDSMultitenantDatabase) validateDatabaseClusterInstallations(installations model.DatabaseClusterInstallations, databaseCluster *model.DatabaseCluster) error {
+	expectedInstallations, err := databaseCluster.GetInstallations()
+	if err != nil {
+		return errors.Errorf("failed to get installations from database cluster ID %s", databaseCluster.ID)
+	}
+
+	if installations.Size() != expectedInstallations.Size() {
+		return errors.Errorf("supplied %s installations, but the cluster ID %s has %s", installations.Size(), expectedInstallations.Size(), databaseCluster.ID)
+	}
+
+	for _, installation := range installations {
+		if !expectedInstallations.Contains(installation) {
+			return errors.Errorf("unable to find installation ID %s in cluster ID %s", installation, databaseCluster.ID)
+		}
+	}
+
+	return nil
+}
+
+func (d *RDSMultitenantDatabase) connectRDSCluster(schema, endpoint, username, password string) (func(logger log.FieldLogger), error) {
+	// This condition allows injecting our own driver for testing.
+	if d.db == nil {
+		db, err := sql.Open("mysql", RDSMySQLConnString(schema, endpoint, username, password))
+		if err != nil {
+			return nil, errors.Wrap(err, "connecting to RDS DB cluster")
+		}
+
+		d.db = db
+	}
+
+	close := func(logger log.FieldLogger) {
+		err := d.db.Close()
+		if err != nil {
+			logger.WithError(err).Errorf("failed to close database connection to %s", endpoint)
+		}
+	}
+
+	return close, nil
+}
+
+func (d *RDSMultitenantDatabase) createUserIfNotExist(ctx context.Context, username, password string) error {
+	query := fmt.Sprintf("CREATE USER IF NOT EXISTS '%s'@'%s' IDENTIFIED BY '%s'", username, "%", password)
+
+	_, err := d.db.ExecContext(ctx, query)
+	if err != nil {
+		return errors.Wrapf(err, "creating user database user %s", username)
```

Comment:
I think the first instance of `user` in this error message was unintentional. Probably should read `"creating database user %s"`